[DRAFT] Rust Foundation: December 2023 Update (#2)
JoelMarcey/alpha-omega: alpha/engagements/2023/rust/update-2023-12.md

Also, Typomania has been adjusted, based on a recent campaign of malicious crates, to make it more sensitive to actual patterns we’re seeing in the wild by adding support for checking if common suffixes have been added or removed from a crate name (e.g., `-rs`).

### Sandpit

Walter has developed a tool called Sandpit as part of the security initiative's Rust ecosystem scanning process. Sandpit scans available crates to try to detect malicious crates based on available heuristics. Sandpit also caught a potentially malicious crate. Sandpit is still under development and there are plans to make it public.

### Disabling Non-canonical Crate Downloads

> Last month, Tobias [announced](https://blog.rust-lang.org/2023/10/27/crates-io-non-canonical-downloads.html) that support for non-canonical downloads on crates.io would be removed. A non-canonical download is the ability to download a crate if you, for example, replace a hyphen in a name with an underscore - crates.io would just normalize those to be the same. Having non-canonical downloads prevented crates.io from using CDNs directly for download requests because of required database lookups that were required for this feature. Now that this is [disabled](https://github.com/rust-lang/crates.io/pull/7549), you have increased performance but, just as important, reliability, since the exact named crate being asked for is always the crate that is attempted to be downloaded.

Reviewer: ```suggestion
Last month, Tobias [announced](https://blog.rust-lang.org/2023/10/27/crates-io-non-canonical-downloads.html) that support for non-canonical downloads on crates.io would be removed. A non-canonical download is the ability to download a crate if you, for example, replace a hyphen in a name with an underscore - crates.io would just normalize those to be the same. Removing support for non-canonical downloads from crates.io allows using CDNs directly for download requests because of required database lookups that were required for supporting this feature. Non-canonical downloads have now been [disabled](https://github.com/rust-lang/crates.io/pull/7549) and the supporting code will be removed next month.
```

Let's not mention the performance and reliability yet, because that will only be true once we implemented download counting for the CDNs.

### Using `async/await` in crates.io

> crates.io code is being ported to using [`async/await`](https://rust-lang.github.io/async-book/01_getting_started/04_async_await_primer.html) in order to remove similar code from packages that are currently unmaintained.

Reviewer: ```suggestion
The crates.io background worker code is being ported to using more [`async/await`](https://rust-lang.github.io/async-book/01_getting_started/04_async_await_primer.html) in order to replace some older dependencies that are no longer unmaintained.
```

### Cargo Version Usage Visualization

> Tobias and JD created a public [dashboard](https://p.datadoghq.com/sb/3a172e20-e9e1-11ed-80e3-da7ad0900002-973f4c1011257befa8598303217bfe3a?refresh_mode=sliding&from_ts=1700688567286&to_ts=1701293367286&live=true) with DataDog that shows how many crates.io download requests are received based on a given `cargo` version. This allows the visualization of how with every release, the amount of crate download requests continue to increase, showing that the crates ecosystem is becoming ever more popular and thus more ripe for potential malicious attacks. 

Reviewer: ```suggestion
Tobias and JD created a public DataDog [dashboard](https://p.datadoghq.com/sb/3a172e20-e9e1-11ed-80e3-da7ad0900002-973f4c1011257befa8598303217bfe3a) that shows how many requests crates.io received per `cargo` version. This allows the visualization of how with every release, the amount of crate download requests continues to increase, showing that the crates ecosystem is becoming ever more popular and thus more ripe for potential malicious attacks. 
```

## Community

Microsoft has [announced](https://x.com/dwizzzleMSFT/status/1720134540822520268?s=20) a further $1M investment into the Rust Foundation.

> The [RFC](https://github.com/rust-lang/rfcs/pull/3463) proposed to [update](https://blog.rust-lang.org/2023/09/22/crates-io-usage-policy-rfc.html) the Terms of Use for crates.io has been approved, merged and landed. It has officially been integrated into the crates.io policy. This RFC is based on prior work by other package managers such as PyPI and npm. The main motivation for this RFC is to give the crates.io team a fixed set of rules to determine if a project is using crates.io in a reasonable way, or if the user should get a warning and the project potentially be removed. It is mostly codifying the existing practices of the team, except for being more strict regarding name squatting. Tobias is currently looking at 20 obvious spam accounts and will most likely remove those crates. For those that may be legitimate, emails will be sent with a grace period.

Reviewer: ```suggestion
The [RFC](https://github.com/rust-lang/rfcs/pull/3463) proposing an [update](https://blog.rust-lang.org/2023/09/22/crates-io-usage-policy-rfc.html) to the Terms of Use for crates.io has been approved, merged and landed. It has officially been integrated into the crates.io policy pages. This RFC is based on prior work by other package registries such as PyPI and npm. The main motivation for this RFC is to give the crates.io team a fixed set of rules to determine if a project is using crates.io in a reasonable way, or if the user should get a warning and the project potentially be removed. It is mostly codifying the existing practices of the team, except for being more strict regarding name squatting. Tobias is currently looking at about 20 obvious spam accounts and will most likely remove those crates. For those that may be legitimate, emails will be sent with a grace period.
```
